using one websocket per exchange connection

barrysteyn opened the pull request:
This will use one websocket per exchange connection. Derived feeds need to be altered to refer to the class's websocket (as opposed to the instance's websocket). This can probably be elegantly changed to by using @property (which would then need no change for derived feeds)

To test that this works, test it on Poloniex by doing the following:
`coin listen -f poloniex -a ETH,XMR -c BTC collect run`
If this is working as per expectation, I will neaten things and put more effort into things.

barrysteyn committed:
using one websocket per exchange connection

barrysteyn commented on numismatic/feeds/base.py:
```diff
@@ -72,7 +72,7 @@ def _rest_client_validator(self, attribute, value):
     def _websocket_client_validator(self, attribute, value):
         self.websocket_client = None if self._websocket_client_class is None \
-            else self._websocket_client_class(websocket=self.websocket)
+            else self._websocket_client_class()
```
These vallidators seem very "hacky" to me. I would like to remove them

barrysteyn replied:
I still reiterate this point. Neither `attribute` nor `value` is being used, and so this must be accomplishing something that it is not designed to do.

snth replied:
It's one method that `attrs` provides of initialising an attribute based on other attributes. I agree that the term *validator* isn't right but if you look beyond that then I think it is fine.

barrysteyn commented on numismatic/feeds/base.py:
```diff
@@ -271,13 +292,15 @@ def listen(self, symbol, channel=None, websocket_url=None):
+                packet = await WebsocketClient.websocket.recv()
                 subscription.raw_stream.emit(packet)
+            except websockets.exceptions.ConnectionClosed:
+                self._connect()
```
need an `await` statement here

barrysteyn commented on numismatic/feeds/base.py:
```diff
@@ -241,15 +239,38 @@ def _make_request(self, api_url, params=None, headers=None, raw=False):
+        if websocket_url not in WebsocketClient.semaphores:
+           WebsocketClient.semaphores[websocket_url] = asyncio.BoundedSemaphore(1)
+
+    async def _connect(self):
```
Note: this is not using the client attribute of subscription. Suggestion: We remove the client attribute of subscription, it is making things more complex than is necessary.

barrysteyn commented on numismatic/feeds/poloniex.py:
```diff
@@ -35,7 +35,7 @@ class PoloniexWebsocketClient(WebsocketClient):
         packet = json.dumps(connection_message)
         logger.info(packet)
-        await self.websocket.send(packet)
+        await PoloniexWebsocketClient.websocket.send(packet)
```
Wuth `@property` on the base, we can get away using `self.websocket` while still adhering to the singleton. I suppose if you like (and agree) with my method, then I will implement that and everything should work

barrysteyn commented on numismatic/feeds/base.py:
```diff
@@ -241,15 +238,56 @@ def _make_request(self, api_url, params=None, headers=None, raw=False):
     exchange = None
     websocket_url = None
-    websocket = attr.ib(default=None)
+    _websocket_map = dict()
```
perhaps this should be called `websocket_connection_pool`

snth replied:
No, map is better. Pool would imply that their substitutable which isn't the case.

barrysteyn commented on numismatic/feeds/base.py:
```diff
@@ -241,15 +238,56 @@ def _make_request(self, api_url, params=None, headers=None, raw=False):
+    _semaphore_map = dict()
+
+    @property
+    def websocket(self):
```
Properties make referring to `websocket` and `semaphore` elegant and easy

snth commented on numismatic/feeds/base.py:
```diff
@@ -241,15 +238,56 @@ def _make_request(self, api_url, params=None, headers=None, raw=False):
+        try:
+            return WebsocketClient._semaphore_map[self.websocket_url]
+        except KeyError:
+            semaphore = asyncio.BoundedSemaphore(1)
```
I see why the semaphores are necessary but I think this makes the code more complicated than I would like it to be. I want to try and refactor this at some point.

snth commented on numismatic/feeds/luno.py:
```diff
@@ -42,13 +42,9 @@ class LunoWebsocketClient(WebsocketClient):
-        return self.websocket
+    def listen(self, symbol, channel=None, websocket_url=None):
+        self.websocket_url = f'{self.websocket_url}/{symbol}'
+        return super().listen(symbol, channel, websocket_url)
```
I don't like the fact that `self.websocket_url` is being set in order to pass the info to `_connect()` while the `websocket_url` parameter is unused. I know the `websocket_url` was being ignored in my code already, just saying. The bigger problem in my mind is that `self.websocket_url` is being set by one subscription and then overwritten by another one later on. In my mind if it's not going to be consistent across subscriptions then it shouldn't be state on the instance and should rather be passed as a parameter.

barrysteyn replied:
We can get rid of `websocket_url`... your concern that `listen` function resets `self.websocket_url` is valid, however I am trying to adapt to your infrastructure and change as little as possible. Perhaps we should have a rethink from the basics... However, coming back to the original question, there is a problem: What (besides url) can we use to uniquely identify a connection? For Luno, their design kinda forces one to do this. Perhaps for luno, we should use `self.base_url` and reset the `self.websocket_url` upon listen? This really sucks, but Luno is forcing this...

barrysteyn committed:
one websocket per unique URL

snth committed:
Makes subscriptions compatible with single websocket connection

snth commented on numismatic/cli.py:
```diff
@@ -274,8 +274,10 @@ def run(state, timeout):
+    # We have one _listener() per client
+    # tasks = {id(sub.client):asyncio.Task(sub.client.listener) 
+    #          for name, sub in state['subscriptions'].items()}
+    tasks = {}
```
The _listener()-s are now started at instance creation time with asyncio.ensure_future() so this is not necessary anymore. I'm thinking about how to remove the whole `run` command as @barrysteyn  requested.

snth commented on numismatic/feeds/base.py:
```diff
@@ -39,7 +39,6 @@ class Subscription:
     symbol = attr.ib()
     channel = attr.ib()
     client = attr.ib()
-    listener = attr.ib()
```
We don't need the `listener` attribute for the Websocket subscriptions anymore but this obviously breaks the REST subscriptions for now.

barrysteyn replied:
Is it wise to merge something that will break existing functionality?

snth replied:
Fixed in latest commit

snth commented on numismatic/feeds/base.py:
```diff
@@ -49,6 +48,10 @@ class Subscription:
+    async def start(self):
+        logger.info(f'Starting Subscription {self.market_name!r} ...')
+        await self.client._subscribe(self)
+
```
Allows subscription to be started as soon as it is created with `asyncio.ensure_future()`.

snth commented on numismatic/feeds/base.py:
```diff
@@ -102,8 +105,9 @@ def subscribe(self, assets, currencies, channels, exchange=None,
-                subscription = websocket_client.listen(symbol, channel)
+                if self.websocket_client is None:
+                    self.websocket_client = self._websocket_client_class()
+                subscription = self.websocket_client.listen(symbol, channel)
```
This was causing multiple clients to be created and as a result multipe websockets. I think this needs to be overhauled because Luno will need to override this to handle their setup of multiple websockets.

barrysteyn replied:
Luno seems weird... however, I thought a middle ground would be to create a new ws client instance for every connection, as long as each client instance shares a unique WS connection that is mapped to their URL

snth replied:
Yes exactly. That is what the code is doing now. I think that seems like the most sensible thing. A `WebsocketClient` is responsible for maintaining a single websocket connection and any subscriptions using that.

snth commented on numismatic/feeds/base.py:
```diff
@@ -245,52 +249,33 @@ class WebsocketClient(abc.ABC):
+    def __attrs_post_init__(self):
+        asyncio.ensure_future(self._listener())
+
+    async def _connect(self, websocket_url=None):
```
I've removed the maps because I think a single `WebsocketClient` should be responsible for a single `Websocket`. Luno's use of multiple websockets should be handled in their feed.

snth commented on numismatic/feeds/base.py:
```diff
@@ -245,52 +249,33 @@ class WebsocketClient(abc.ABC):
-            websocket = await websockets.connect(websocket_url)
-            self._websocket_map[websocket_url] = websocket
-            if hasattr(self, 'on_connect'):
-                await self.on_connect(WebsocketClient.websocket)
```
I think `on_connect` was only being used by Bitfinex and that use case could be handled by the existing message handler mechanism so I removed this special case.

snth commented on numismatic/feeds/base.py:
```diff
@@ -245,52 +249,33 @@ class WebsocketClient(abc.ABC):
+            Connects to websocket. Uses a future to ensure that only one
+            connection at a time will happen
         '''
-        with (await self.semaphore):
```
As said previously, I didn't like introducing semaphores because it introduces another concept. Asyncio documentation doesn't usually talk about them. Async programmers are more familiar with Futures and Tasks so I prefer using those concepts to solve the problem here.

snth commented on numismatic/feeds/base.py:
```diff
@@ -351,7 +323,17 @@ def __handle_packet(packet, subscription):
+                if result is STOP_HANDLERS:
+                    # FIXME: should this raw_stream now rather sit on the
+                    # WebsocketClient instead of the Subscription?
+                    subscription.raw_stream.emit(packet)
```
The raw `packet`s now still go to the subscriptions but I think they should really be associated with the websocket_client. So we need to move the `raw_stream` from the subscription to the websocket_client. I think that's probably best handled in a separate PR though.

barrysteyn replied:
Yes I agree. In general, it seems like Subscription is trying to do too much.

snth replied:
Beyond that, I see the main purpose of capturing the raw packets for debugging. As such they should be captured as close to the source as possible. By the time they hit a subscription they've already been filtered and assigned to a subscription. If that logic is faulty then having the raw packet end up in the wrong place isn't really going to help with the debugging. Also, you might lose the sequencing implicit in the packet arrival this way which is another downside.

I'll do this refactoring in separate PR.

snth committed:
Simplifies WebsocketClient._connect()

snth committed:
Introduces blocking WebsocketClient construction which ensures websocket availability

The idea is that a WebsocketClient should be fully constructed when it is
handed over.

snth committed:
Makes WebsocketClient construction asynchronous again

snth committed:
Cleans up the CLI run command

snth commented:
@barrysteyn I tried removing the `run` command but the part after `coin()` in the `__main__` section never gets run and we need somewhere to run `loop.run_forever()`.

barrysteyn commented on numismatic/feeds/base.py:
```diff
@@ -241,18 +242,39 @@ def _make_request(self, api_url, params=None, headers=None, raw=False):
+    websocket = attr.ib(default=None, repr=False)
+    subscriptions = attr.ib(default=attr.Factory(list), repr=False)
+
+    def __attrs_post_init__(self):
```
Nice. :+1:

barrysteyn commented on numismatic/feeds/base.py:
```diff
@@ -241,18 +242,39 @@ def _make_request(self, api_url, params=None, headers=None, raw=False):
+        asyncio.ensure_future(self._connect())
+        asyncio.ensure_future(self._listener())
+
+    async def _connect(self):
```
This does seem much neater, well done :+1:

barrysteyn commented on numismatic/feeds/base.py:
```diff
@@ -241,18 +242,39 @@ def _make_request(self, api_url, params=None, headers=None, raw=False):
+                asyncio.ensure_future(websockets.connect(self.websocket_url))
+        if isinstance(self.websocket, asyncio.Future):
+            self.websocket = await self.websocket
+        return self.websocket
```
I don't understand this pattern: A side-effect is created by setting the instance (`self.websocket`) and in addition, `self.websocket` is returned.

snth replied:
You're right, it shouldn't return the websocket anymore. I was trying to avoid having a method with a side-effect as I don't like them and try to avoid them wherever possible. One of my intermediate commits did achieve this and that code was much neater in my opinion. However it sacrificed the concurrency in connecting to the websocket. I think that's an acceptable cost and won't make any difference in production use cases but it made startup slower in interactive sessions in the shell so I opted against it for now. We could always bring it back again later.

barrysteyn commented on numismatic/feeds/base.py:
```diff
@@ -261,55 +283,37 @@ def listen(self, symbol, channel=None, websocket_url=None):
-            self.websocket = await self._connect(subscription)
-        await self._subscribe(subscription)
+    async def _listener(self):
+        await self._connect()
```
Regarding connect, I have an idea. Lets create a connection class within the webocket_client class. We will implement the dunders `__enter__` and `__exit__`, and then we can do this:
```
with self.connection():
  while True:
    ...
```
Even the `ConnectionClosed` exception can be handled well..

snth committed:
Fixes case where we need to reconnect to a closed websocket

barrysteyn committed:
pololniex using channel map for subscriptions

snth committed:
Fixes RestClient subscriptions to work with new Subscription setup

snth committed:
Makes it clearer that _connect() acts through side-effects

The previous code didn't make this clear as pointed out by @barrysteyn
